### ISSUE TITLE
Lexer: Use enums instead of sets of `const`s

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#26951]
#26919 introduced various sets of `u8` consts which represent various states in the `disambiguate` code e.g.:

```rs
pub(super) const D_NONE: u8 = 0;
pub(super) const D_BINDING: u8 = 1;
pub(super) const D_BOUND: u8 = 2;
pub(super) const D_INIT: u8 = 3;
```

https://github.com/oxc-project/oxc/blob/591966d8dd5eb7ee69be437c89e4fa40076b5a3d/crates/oxc_lexer/src/pipeline/disambiguate/context/frame.rs#L99-L193

In Rust this is not considered idiomatic - it's preferable to use enums for type-safety.

```rs
#[repr(u8)]
pub(super) enum Declarator {
    None = 0,
    Binding = 1,
    Bound = 2,
    Init = 3,
}
```

Additionally, it appears that the lack of types has disguised some unreachable code. If this code used `enum`s, Clippy would warn that certain values are never read.

It's unclear to me if it's just dead code which can be removed, or the code *should* be reachable and there are logic bugs resulting in it not being. Below is what Claude said about it (I can't really evaluate how much truth in what he says):

---

The `context/` walker keeps its per-frame state in untyped `u8` fields whose meaning depends on `Frame.kind`: ten constant groups (48 constants) share four fields on `Frame`. Nothing ties a group to the frame kind it belongs to, so the invariants that keep the groups apart are held only by convention and by comments.

Auditing those invariants turned up three values that cannot do anything, one field that is never read, and two comments that describe states the code never produces. None of them is a lexing bug - every case is unreachable rather than wrong - but each took tracing several call sites across three files to establish, and none would have been expressible had the fields been typed.

Evidence below is from `fda9845370`; line numbers are as of that commit.

## Which group lives in which field

| Field       | Groups it carries                                                                 | Frame kinds                                                                                    |
| ----------- | --------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------- |
| `state: u8` | `D_*`, `F_*`, `M_*`, `R_*`, `A_*`, `C_EXTENDS`/`C_IMPLEMENTS`, `L_INTERFACE_BODY` | statement holders, `Head`, `Object`/`ClassBody`, `TypeRegion`, `Angle`, `ClassHead`, `TypeLit` |
| `reg: u8`   | `S_*`, `C_INTERFACE`                                                              | statement holders, `ClassHead`                                                                 |
| `head: u8`  | `H_*`, plus an unnamed `0` for "none"                                             | statement holders, `Head`                                                                      |
| `mods: u8`  | `MOD_*`                                                                           | `Object`, `ClassBody`                                                                          |

Collisions between groups are safe only because of invariants that are nowhere stated in the types:
`C_EXTENDS` (1) and `C_INTERFACE` (1) are different fields; `M_KEY_SEEN` (1) and `L_INTERFACE_BODY` (1) are different frame kinds; `D_BINDING` (1) and `F_BOUND` (1) are different frame kinds.

## 1. `Frame.decl_binding` is never read, and its three writers are unreachable

`decl_binding` is declared in `context/frame.rs:174`:

```rust
/// For-head: the binding came from a declaration (`for (let x of`).
pub(super) decl_binding: bool,
```

It is set `true` at three sites — `context/punct.rs:585`, `context/punct.rs:776` and `context/words.rs:493` —
each of them in this shape:

```rust
if let Some(di) = self.decl_frame() {
    self.frames[di].state = D_BOUND;
    if self.frames[di].kind == FrameKind::Head {
        self.frames[di].decl_binding = true;
    }
}
```

`decl_frame()` (`context/walker.rs:239-242`) returns `Some(i)` only when the frame is a statement holder:

```rust
pub(super) fn decl_frame(&self) -> Option<usize> {
    let i = self.stmt_frame();
    if is_stmt_holder(self.frames[i].kind) { Some(i) } else { None }
}
```

and `is_stmt_holder` (`context/frame.rs:212-221`) is `Root | Block | FnBody | ArrowBody | StaticBlock`.

`Head` is not among them, so all three `kind == FrameKind::Head` tests are always false, and the field is never set. It is not read anywhere either — the only other mentions are the three places that clear it (`context/punct.rs:386`, `context/punct.rs:691`, `context/frame.rs:203`, `context/walker.rs:149`).

The same reasoning makes `sf.kind != FrameKind::Head` at `context/walker.rs:614` always true.

Worth confirming: was `for (let x of` meant to be distinguished from `for (x of`? If so, the distinction is not currently being made anywhere. If not, the field and its three writers can go.

## 2. `C_IMPLEMENTS` is written but never read

`C_IMPLEMENTS` (`context/frame.rs:148`) has exactly one mention outside its definition, `context/words.rs:320-324`:

```rust
tk!(KwImplements) => {
    // Type references follow.
    self.top_mut().state = C_IMPLEMENTS;
    self.keyword(tk!(KwImplements));
}
```

Nothing reads it. Its only effect is to overwrite `C_EXTENDS` on the `ClassHead`, and that cannot change behaviour either: the sole reader of `C_EXTENDS` (`context/punct.rs:458`) also requires `self.prev_kw == tk!(KwExtends)`, which cannot hold once `implements` has been processed.

Note also that the write is unguarded, unlike the `C_EXTENDS` write two arms above it, which checks `self.top_kind() == FrameKind::ClassHead` first. It is safe today only because `resolve_keyword` turns `implements` into a plain name unless a `ClassHead` is on top.

## 3. `R_INTERFACE` is read but never written

`R_INTERFACE` (`context/frame.rs:138`) is read at three sites and written at none.

`open_region` (`context/types.rs:9-16`) is the only `push(FrameKind::TypeRegion)` in the module, and every one of its twelve call sites passes a literal: `R_STMT` ×2, `R_INLINE` ×6, `R_ARROW_RET` ×1, `R_EXPR` ×2, `R_ASSERT` ×1. No region is ever given `R_INTERFACE`, so:

- `context/types.rs:95-108`, the `interface X extends Y {` body branch, is unreachable. Interface bodies are in fact produced by `context/punct.rs:446-455`, which keys off `reg == C_INTERFACE` on the `ClassHead` instead.
- `context/step.rs:263`, `R_STMT | R_INTERFACE => self.end_statement()`, is equivalent to `R_STMT =>`.
- `context/walker.rs:591`, `matches!(r.state, R_INLINE | R_STMT | R_INTERFACE)`, is equivalent to `matches!(r.state, R_INLINE | R_STMT)`.

## 4. Two comments describe states that never occur

Because the fields are untyped, these comments are the only documentation of which group owns which field - they are doing the job a type would do, and both are wrong.

`context/frame.rs:71`:

```rust
// Declarator state on statement frames and for-heads (`state`).
```

`D_*` never reaches a for-head. Every `D_*` read and write goes through `decl_frame()`, which excludes `Head` (see §1). A `Head` frame's `state` holds `F_*` and nothing else.

`context/frame.rs:118`:

```rust
// Member state on Object / ClassBody / TypeLit (`state`).
```

`M_*` never reaches a `TypeLit`. All twelve `state = M_*` writes are guarded to `Object` / `ClassBody` (`context/punct.rs` 97, 268, 306, 393, 413, 769; `context/step.rs` 237, 269, 348; `context/walker.rs` 511, 712; `context/words.rs` 566, guarded by its caller at `context/words.rs:45`). A `TypeLit`'s `state` only ever holds `L_INTERFACE_BODY`.

## Suggested fix: give the groups types

Seven of the ten groups convert to plain enums with no restructuring. Every use in them is `==`, `!=`, `matches!`, or a match arm with a `_` fallback - there is no arithmetic, ordering or bit manipulation anywhere in the seven:

- `S_*` (14 values, 63 sites) is the largest and the least invasive: every read and write already goes through `stmt_reg()` / `set_stmt_reg()` (`context/walker.rs:245-253`), so only those two signatures change. `S_NONE` becomes the `Default`. The unused discriminant 10 disappears with the constants.
- `R_*` (6): `open_region(rule: u8, decl: bool)` becomes `open_region(rule: RegionRule, decl: bool)`.
- `A_*` (4), `D_*` (4), `F_*` (4), `M_*` (3): each confined to one frame kind, as established above.
- `H_*` (6): `head` carries nothing else, and its `0` means "no pending head keyword", so `Option<HeadKind>` is exact - `head != 0` becomes `.is_some()`.

Two details that fall out of the audit: `R_*` and `A_*` never hold `0` (every `TypeRegion` and every `Angle` is given a state at its push site), so neither needs a "none" variant; and `C_INTERFACE` and `L_INTERFACE_BODY` are single-valued groups, i.e. booleans rather than enums.

`MOD_*` should stay as it is — the three values are bit flags (1, 2, 4), combined with `|=` and tested with `& != 0`.

That leaves the `ClassHead` corner (`C_*` over two fields, one of which, `reg`, it shares with `S_*`) as the only part needing a judgement call rather than a mechanical change.

Two smaller things in the same struct, for whoever is in there: `prologue: u8` only ever holds 0 or 1, and `open_questions: u16` is a counter rather than a state.
